### ISSUE TITLE
suggestion to add function for changing models without deleting/creating new translators

### ISSUE THREAD

**Issue by panosk** [OpenNMT/CTranslate2#76]
Hi @guillaumekln ,

As far as I can see, if we create an instance of `Translator`, we can't change the model without destroying the object and creating a new one, as the model can only be defined in the constructor --unless I missed sth. Wouldn't it make sense to have a function to change the current model? Even if `delete`ing and making `new` translators is trivial, IMO it would improve the already excellent interface. If this makes sense, I could work on that soon, when I have some time.

**Comment by guillaumekln** [comment]
Hi,

Just to make sure, are you referring to this class https://github.com/OpenNMT/CTranslate2/blob/master/include/ctranslate2/translator.h? If yes, this sounds like a valid use case.

**Comment by panosk** [comment]
Yes, this class is the user interface for actually using the library, and that's why I'm suggesting this. Right now when I want to change the model in an application I have to do this
```
if (translator) {
        delete translator;
        translator = nullptr;
    }

    translator = new ctranslate2::Translator(new_model_path,
                                             ctranslate2::Device::CPU);
```

**Comment by guillaumekln** [comment]
You could also use a `std::unique_ptr` and call reset. ;)

**Comment by panosk** [comment]
Indeed that would save a few lines of code, but I was thinking to add a setter function for the `_model` shared_ptr member and reset that instead :).

**Comment by guillaumekln** [comment]
Yeah I would accept a PR for that.

**Comment by panosk** [comment]
Hi @guillaumekln 

I came up with this which seems to work fine (after removing the `const`ness from the `_model` of course):
```
void Translator::set_model(const std::string &model_dir)
  {
      std::shared_ptr<const models::Model> model;
      _model = model->load(model_dir,this->device());
      make_graph();
  }
```
I digged as much as I could into the code but I'd like your insight: I think this solution also requires to define getter functions for `_device_index` and `_compute_type` in `model.h` so we keep everything as is and only change the model. Otherwise, the assignment to the new model will change these values to their defaults. Is that the case?

**Comment by guillaumekln** [comment]
Hi,

In general, a `set_model` method should allow resetting this properties to new values. But for convenience, it's looks OK to have an overload that just takes the `model_dir` and reuse the same device and compute type. You should indeed add getters for these properties and comment this behavior in the `Translator` interface.

I suggest also adding a more general overload such as `set_model(const std::shared<const models::Model>& model)`.

**Comment by panosk** [comment]
All good points, thank you. I will send a PR in the next couple of days.

**Comment by guillaumekln** [comment]
Closed by #84.
